File: labs/control-ui/api-shadow/subscription_crawler_pipeline.ts
export interface CrawlSource {
  url: string;
  intervalSecs: number;
  lastCrawl: number;
  totalHarvested: number;
  enabled: boolean;
}

interface NodeBufferShim {
  Buffer?: {
    from(input: string, encoding: string): { toString(encoding: string): string };
  };
}

function nodeGlobalBuffer() {
  return (globalThis as NodeBufferShim).Buffer;
}

function decodeBase64Safe(input: string): string {
  try {
    if (typeof atob === 'function') {
      return atob(input);
    }
    const buf = nodeGlobalBuffer();
    if (buf) {
      return buf.from(input, 'base64').toString('utf-8');
    }
  } catch {
    // Return original string if base64 decoding fails
  }
  return input;
}

export class SubscriptionCrawlerPipeline {
  private sources: Map<string, CrawlSource> = new Map();
  private harvestedProxies: Set<string> = new Set();

  addSource(url: string, intervalSecs: number = 3600): void {
    this.sources.set(url, {
      url,
      intervalSecs: Math.max(60, intervalSecs),
      lastCrawl: 0,
      totalHarvested: 0,
      enabled: true,
    });
  }

  dispatchPendingSources(currentTime: number): string[] {
    const toCrawl: string[] = [];
    for (const source of this.sources.values()) {
      if (
        source.enabled &&
        (source.lastCrawl === 0 || currentTime >= source.lastCrawl + source.intervalSecs)
      ) {
        source.lastCrawl = currentTime;
        toCrawl.push(source.url);
      }
    }
    return toCrawl;
  }

  ingestCrawlContent(sourceUrl: string, rawContent: string): number {
    let count = 0;
    const contentTrimmed = rawContent.trim();

    let decodedText = contentTrimmed;
    const possibleDecoded = decodeBase64Safe(contentTrimmed);
    if (possibleDecoded !== contentTrimmed && (
      possibleDecoded.includes('://') || possibleDecoded.includes('\n')
    )) {
      decodedText = possibleDecoded;
    }

    const lines = decodedText.split('\n');
    for (const line of lines) {
      const trimmed = line.trim();
      if (
        trimmed.startsWith('ss://') ||
        trimmed.startsWith('vmess://') ||
        trimmed.startsWith('vless://') ||
        trimmed.startsWith('trojan://') ||
        trimmed.startsWith('hysteria2://') ||
        trimmed.startsWith('tuic://')
      ) {
        if (!this.harvestedProxies.has(trimmed)) {
          this.harvestedProxies.add(trimmed);
          count++;
        }
      }
    }

    const source = this.sources.get(sourceUrl);
    if (source) {
      source.totalHarvested += count;
    }

    return count;
  }

  getHarvestedProxies(): string[] {
    return Array.from(this.harvestedProxies).sort();
  }

  totalHarvestedCount(): number {
    return this.harvestedProxies.size;
  }
}

File: labs/mobile/donor-archive/tunnel/SubscriptionCrawlerPipeline.kt
package com.luminet.android.tunnel

import java.util.Base64
import java.util.HashSet

data class CrawlSource(
    val url: String,
    val intervalSecs: Long,
    var lastCrawl: Long = 0,
    var totalHarvested: Int = 0,
    var enabled: Boolean = true
)

class SubscriptionCrawlerPipeline {
    private val sources = HashMap<String, CrawlSource>()
    private val harvested = HashSet<String>()

    fun addSource(url: String, intervalSecs: Long) {
        val interval = intervalSecs.coerceAtLeast(60)
        sources[url] = CrawlSource(url = url, intervalSecs = interval)
    }

    fun dispatchPendingSources(currentTime: Long): List<String> {
        val toCrawl = ArrayList<String>()
        for (src in sources.values) {
            if (src.enabled && (src.lastCrawl == 0L || currentTime >= src.lastCrawl + src.intervalSecs)) {
                src.lastCrawl = currentTime
                toCrawl.add(src.url)
            }
        }
        return toCrawl
    }

    fun ingestCrawlContent(sourceUrl: String, rawContent: String): Int {
        val trimmed = rawContent.trim()
        val decodedText = try {
            val decoded = Base64.getDecoder().decode(trimmed)
            String(decoded, Charsets.UTF_8)
        } catch (_: Exception) {
            trimmed
        }

        var count = 0
        for (line in decodedText.lines()) {
            val l = line.trim()
            if (l.startsWith("ss://") ||
                l.startsWith("vmess://") ||
                l.startsWith("vless://") ||
                l.startsWith("trojan://") ||
                l.startsWith("hysteria2://") ||
                l.startsWith("tuic://")
            ) {
                if (harvested.add(l)) {
                    count++
                }
            }
        }

        sources[sourceUrl]?.let {
            it.totalHarvested += count
        }

        return count
    }

    fun getHarvestedProxies(): List<String> = harvested.sorted()

    fun totalHarvestedCount(): Int = harvested.size
}
